apm-agent-core/src/main/java/co/elastic/apm/agent/metrics/builtin/AgentOverheadMetrics.java

```diff
@@ -0,0 +1,453 @@
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package co.elastic.apm.agent.metrics.builtin;
+
+import co.elastic.apm.agent.configuration.MetricsConfiguration;
+import co.elastic.apm.agent.context.AbstractLifecycleListener;
+import co.elastic.apm.agent.impl.ElasticApmTracer;
+import co.elastic.apm.agent.metrics.Labels;
+import co.elastic.apm.agent.metrics.MetricCollector;
+import co.elastic.apm.agent.metrics.MetricRegistry;
+import co.elastic.apm.agent.metrics.MetricsProvider;
+import co.elastic.apm.agent.util.ElasticThreadStateListener;
+import co.elastic.apm.agent.util.ExecutorUtils;
+import co.elastic.apm.agent.util.JmxUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import java.lang.management.ManagementFactory;
+import java.lang.management.OperatingSystemMXBean;
+import java.lang.management.ThreadMXBean;
+import java.lang.reflect.Method;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class AgentOverheadMetrics extends AbstractLifecycleListener implements ElasticThreadStateListener, MetricsProvider {
+
+    private static final Logger logger = LoggerFactory.getLogger(AgentOverheadMetrics.class);
+    private static final String CPU_OVERHEAD_METRIC = "agent.background.cpu.overhead.pct";
+    private static final String CPU_USAGE_METRIC = "agent.background.cpu.total.pct";
+    private static final String ALLOCATION_METRIC = "agent.background.memory.allocation.bytes";
+    private static final String THREAD_COUNT_METRIC = "agent.background.threads.count";
+
+    private boolean cpuOverheadMetricEnabled;
+    private boolean cpuUsageMetricEnabled;
+    private boolean allocationMetricEnabled;
+    private boolean threadCountMetricEnabled;
+
+    @Nullable
+    private Long lastReportedProcessCpuTime;
+
+    /**
+     * Contains all started, running and stopped threads since the last metrics reporting.
+     */
+    private final ConcurrentHashMap<Thread, ThreadInfo> lastThreadInfo = new ConcurrentHashMap<>();
+
+    private static class ThreadInfo {
+        final String threadPurpose;
+
+        @Nullable
+        volatile Long cpuTime;
```

Comment:
Would be better to avoid the `Long` object boxing and unboxing if it's straightforward